pkg/cover/report.go

```diff
@@ -212,14 +225,21 @@ func getFile(files map[string]*file, name, path, module string) *file {
 }
 
 func (rg *ReportGenerator) findSymbol(pc uint64) *backend.Symbol {
-	idx := sort.Search(len(rg.Symbols), func(i int) bool {
-		return pc < rg.Symbols[i].End
+	idx := sort.Search(len(rg.Ranges), func(i int) bool {
+		return pc < rg.Ranges[i].Start
 	})
-	if idx == len(rg.Symbols) {
+	if idx == 0 {
 		return nil
 	}
-	s := rg.Symbols[idx]
-	if pc < s.Start || pc > s.End {
+	idx--
+	var s *backend.Symbol
+	for j := idx; j < len(rg.Ranges); j++ {
```

Comment:
This function was apparently meant to be very fast (therefore there was only a binary search and a map access). With the newly introduced loop such a guarantee no longer exists.

Can you please explain why you needed the loop?

Reply:
the rg.Ranges are sorted here, 1st by End address, then by start address. The idea is to let inline function appear before normal function.
```
	sort.Slice(allRanges, func(i, j int) bool {
		if allRanges[i].End != allRanges[j].End {
			return allRanges[i].End < allRanges[j].End
		}
		return allRanges[i].Start > allRanges[j].Start
	})
```
So the sort.Search only return the idx first satisfy pc >= rg.Ranges[i].Start (which might be an inline function), but we still need to continue the search from idx if the pc is inside normal function.

In average in Android kernel case, there are less than 10 inline functions for each function. So the j loop maximum will iterate 10 times, but if the pc belongs to inline function, the binary search will be enough and the j loop will only iterate once.

I didn't find any better solution than this.

Reply:
```
	idx := sort.Search(len(rg.Ranges), func(i int) bool {
		return pc < rg.Ranges[i].Start
	})
```

This will find the first index when `rg.Ranges[i].Start` is bigger than `pc` (assuming that `rg.Ranges` are sorted by `Start` in an increasing order).

One point to verify -- is the precondition valid? From your sorting loop I see that it's only guaranteed that it's ordered by `End`. The existing code removes overlapping symbols, so if it's sorted by one end, it's also sorted by the other. But the new code, as I see, keeps everything - and therefore it's not correct to assume that we can binary-search over `Start`.

Maybe it makes sense to preprocess the ranges array so that it also only contains non-overlapping segments? But instead of just deleting ones that overlap you can (and should, as it seems) split big ranges into smaller ones, giving priority to those segments that are smaller. So, as a result, in `findSymbol` you will be able to just binary-search the segment that contains the PC and take a symbol pointer from it.

Reply:
I tried to implement as you suggested, but I don't think it can overcome this problem to do binary search once.

Basically it works (split into ranges and sort) if normal function contains only one level inline functions.
However, it won't work if 1st level inline function contains 2nd level inline functions, even further level of inline functions. For this case, we still need to use binary search + linear search.

So I think sort range by End and then by Start => do binary search + limit linear search is still better. I didn't find any performance issue with this for 1 million pcs. The buildSymbol function finishes within 1 second.

Reply:
updated the patch, pls check again.

Reply:
Here is my understanding of your PR and comments, please correct me where I'm wrong. 
1. `allSymbols` in `makeDWARFUnsafe` contains both normal functions and various inlined functions.
2. For each function, we have a number of PC ranges that belong to them.
3. For normal functions, their ranges also include PCs that belong to functions inlined to them.
4. Inlined functions can also have other inlined functions in them, and their PC ranges would also include PC of those inlined-inlined functions.

From the points above it follows that ranges are either completely contained within each other or don't intersect at all. De-facto there's just some kind of tree-like structure behind it.

And in `findSymbol` we either want to find the covering range that has the minimum width (end PC - start PC) and covers the requested PC. Or, in a tree interpretation, we want to each a leaf node.

But also the current approach to do it is fast enough. Because if you find e.g. the first `End` of a range that's bigger than the requested PC, then in the worst case you need to traverse all nested inline functions whose ranges end on the same `End` PC. Which just cannot be a too big number.

Reply:
Basically yes, let me put more info here in case future we or someone need to revisit this.

There are some issues on current way of elf parsing:
1. Functions are reading from .symbtab section and are assumed that the function is in continous address space. But the truth is not. I observed some functions are placed in ranges. For example asan.module_ctor below.
```
0000000000000000 l       .text  0000000000000000 $x.3
0000000000000638 l     F .text  0000000000000024 asan.module_ctor
000000000000065c l     F .text  0000000000000024 asan.module_dtor
0000000000000000 l    d  .text  0000000000000000 .text
0000000000000680 l       .text  0000000000000000 $x.1
0000000000000680 l     F .text  000000000000000c asan.module_ctor
```
3. No inline function info is present and so header files are assumed to have 1 total pc.
4. Functions from .init.text or .exit.text sections are not excluded, the function start address for some symbols can be same. So during findSymbol, it might find incorrect one. (correctly me if it's already considered)

And the PR is to leverage the inline function info from dwarf records.
1. yes for Alex's point 1, allSymbols contains both normal functions and various inlined functions. In dwarf presentation, each normal function (aka subprogram) might contain several inline functions. So for each inline function, we know which normal function it belongs, but we don't know if it's a child of another inline function easily.
2. Yes for Alex's point 2. To avoid the relationship building of normal function and inline function, even inline function of another inline function, we use dwarf pc range (allRanges) to first sort by range.End and then range.Start, so that we can do binary search + linear search later to find which function the pc belongs to in buildSymbols.
3. Yes for Alex's point 3. Normal functions ranges cover their inline function ranges. However, 1 pc either belongs to one normal function or one inline function. In this way, we can later do statics in /cover or /subsystemcover to say how many covered pcs in normal functions, and how many covered pcs in all functions including inline functions.
5. Yes for Alex's point 4. In current PR implementation, wo don't need to walk inline function recursively.
6. There might be a tree structure, but when I decoded the debuginfo, the edge info is missing at least for inline function of another inline function. Also, if there is a tree, then it's tree of DIE while not tree of ranges. So the address layout can have many senorios, for example function A range 1 + function a range 1 + function A range 2 + function a range 2 + function b range1... It seems harder to construct back to a tree.
7. Yes, the current approach is pretty fast, 1 second for 1 million pcs in buildSymbols. So it shouldn't be a problem during findSymbol operations.
8. Meanwhile, only functions in .text section are considered, I think it's more syzkaller friendly. Because kcov process doesn't collect any coverage from functions in .init.text and .exit.text sections.

The overall performance is not too bad, I think. The makeDWARFUnsafe only runs once which takes 1min20s, while existing way of decoding from symbtab takes ~1min if I remember correctly (on normal PC with 4 cores). For blade servers with many cores, I didn't see much impact. For refreshing /cover or /subsystemcover page, they finished in seconds.

Thanks

Reply:
Any further consideration?

Reply:
Sorry for the delay,  I'll take one more look with a fresh mind.